src/libraries/System.Runtime.InteropServices.JavaScript/gen/JsImportGenerator/JSGeneratorFactory.cs

```diff
@@ -0,0 +1,317 @@
+
+                // discard no void
+                case { } when jsMarshalingInfo.JSType == JSTypeFlags.Discard:
+                    throw fail("TODO discard only on void");
+
+                // primitive
+                case { ManagedType: JSSimpleTypeInfo simple }:
+                    return Create(info, isToJs, simple.KnownType, Array.Empty<KnownManagedType>(), jsMarshalingInfo.JSType, Array.Empty<JSTypeFlags>(), fail);
+
+                // nullable
+                case { ManagedType: JSNullableTypeInfo nullable }:
+                    return Create(info, isToJs, nullable.KnownType, new[] { nullable.ResultTypeInfo.KnownType }, jsMarshalingInfo.JSType, null, fail);
+
+                // array
+                case { ManagedType: JSArrayTypeInfo array }:
+                    return Create(info, isToJs, array.KnownType, new[] { array.ElementTypeInfo.KnownType }, jsMarshalingInfo.JSType, jsMarshalingInfo.JSTypeArguments, fail);
+
+                // array segment
+                case { ManagedType: JSArraySegmentTypeInfo segment }:
+                    return Create(info, isToJs, segment.KnownType, new[] { segment.ElementTypeInfo.KnownType }, jsMarshalingInfo.JSType, jsMarshalingInfo.JSTypeArguments, fail);
+
+                // span
+                case { ManagedType: JSSpanTypeInfo span }:
+                    return Create(info, isToJs, span.KnownType, new[] { span.ElementTypeInfo.KnownType }, jsMarshalingInfo.JSType, jsMarshalingInfo.JSTypeArguments, fail);
+
+                // task
+                case { ManagedType: JSTaskTypeInfo task } when task.ResultTypeInfo is JSSimpleTypeInfo taskRes && taskRes.FullTypeName == "void":
+                    return Create(info, isToJs, task.KnownType, Array.Empty<KnownManagedType>(), jsMarshalingInfo.JSType, jsMarshalingInfo.JSTypeArguments, fail);
+                case { ManagedType: JSTaskTypeInfo task }:
+                    return Create(info, isToJs, task.KnownType, new[] { task.ResultTypeInfo.KnownType }, jsMarshalingInfo.JSType, jsMarshalingInfo.JSTypeArguments, fail);
+
+                // action + function
+                case { ManagedType: JSFunctionTypeInfo function }:
+                    return Create(info, isToJs, function.KnownType, function.ArgsTypeInfo.Select(a => a.KnownType).ToArray(), jsMarshalingInfo.JSType, jsMarshalingInfo.JSTypeArguments, fail);
+
+                default:
+                    //throw new MarshallingNotSupportedException(info, context);
+                    throw new NotImplementedException(info.ManagedType.GetType().FullName + ": " + info.ManagedType.FullTypeName + " - " + info.ManagedIndex + " " + info.InstanceIdentifier);
+            }
+        }
+
+        internal static BaseJSGenerator Create(TypePositionInfo info, bool isToJs, KnownManagedType marshaledType, KnownManagedType[] argumentTypes, JSTypeFlags jsType, JSTypeFlags[] jsTypeArguments, Func<string, Exception> failWithReason)
+        {
+            switch (marshaledType)
```

Comment:
My personal preference would be to not have it as a single switch. There are branches that have a lot of conditions duplicated. Wouldn't it be more readable as combination of switches and ifs?

Reply:
Maybe set of separate functions containing smaller switch. Let's do that once we are past preview7 deadline.